argon2/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+
+        if let Some(secret) = &secret {
+            if MAX_SECRET < secret.len() {
+                return Err(Error::SecretTooLong);
+            }
+        }
+
+        // Validate memory cost
+        if MIN_MEMORY > m_cost {
+            return Err(Error::MemoryTooLittle);
+        }
+
+        if MAX_MEMORY < m_cost {
+            return Err(Error::MemoryTooMuch);
+        }
+
+        if m_cost < 8 * lanes {
+            return Err(Error::MemoryTooLittle);
+        }
+
+        // Validate time cost
+        if MIN_TIME > t_cost {
+            return Err(Error::TimeTooSmall);
+        }
+
+        if MAX_TIME < t_cost {
+            return Err(Error::TimeTooLarge);
+        }
+
+        // Validate lanes
+        if MIN_LANES > lanes {
+            return Err(Error::LanesTooFew);
+        }
+
+        if MAX_LANES < parallelism {
+            return Err(Error::LanesTooMany);
+        }
+
+        // Validate threads
+        if MIN_THREADS > lanes {
+            return Err(Error::ThreadsTooFew);
+        }
+
+        if MAX_THREADS < parallelism {
+            return Err(Error::ThreadsTooMany);
+        }
+
+        // Align memory size
+        // Minimum memory_blocks = 8L blocks, where L is the number of lanes
+        let memory_blocks = if m_cost < 2 * SYNC_POINTS * lanes {
+            2 * SYNC_POINTS * lanes
+        } else {
+            m_cost
+        };
+
+        let segment_length = memory_blocks / (lanes * SYNC_POINTS);
+
+        Ok(Self {
+            secret,
+            t_cost,
+            m_cost,
+            lanes,
+            threads: parallelism,
+            segment_length,
+            version,
+        })
+    }
+
+    /// Function that performs memory-hard hashing with certain degree of parallelism.
+    pub fn hash_password(
+        &self,
+        argon2_type: Algorithm,
+        pwd: &[u8],
+        salt: &[u8],
+        ad: &[u8],
+        out: &mut [u8],
+    ) -> Result<()> {
+        // Validate output length
+        if MIN_OUTLEN > out.len() {
+            return Err(Error::OutputTooShort);
+        }
+
+        if MAX_OUTLEN < out.len() {
+            return Err(Error::OutputTooLong);
+        }
+
+        if MAX_PWD_LENGTH < pwd.len() {
+            return Err(Error::PwdTooLong);
+        }
+
+        // Validate salt (required param)
+        if MIN_SALT_LENGTH > salt.len() {
+            return Err(Error::SaltTooShort);
+        }
+
+        if MAX_SALT_LENGTH < salt.len() {
+            return Err(Error::SaltTooLong);
+        }
+
+        // Validate associated data (optional param)
+        if MAX_AD_LENGTH < ad.len() {
+            return Err(Error::AdTooLong);
+        }
+
+        let memory_blocks = (self.segment_length * self.lanes * SYNC_POINTS) as usize;
+
+        // Hashing all inputs
+        #[allow(unused_mut)]
+        let mut initial_hash = self.initial_hash(argon2_type, pwd, salt, ad, out);
+        let mut memory = vec![Block::default(); memory_blocks];
```

Comment:
This is the only part of the code that relies on `alloc`.

It'd be pretty easy to extract a method that operates over a `&mut [Block]` (or a newtype wrapping it), but it'd probably be nice to encapsulate it instead of exposing `Block`, and if we do that, waiting for const generics would be nice.

So I left it out of the initial implementation, but it should be pretty easy to add retroactively.